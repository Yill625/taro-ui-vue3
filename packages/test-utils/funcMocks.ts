import * as utils from "@taro-ui-vue3/utils/common"
import { pxTransformMockFn, getEnvMockFn } from './@tarojs/taro'

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let pxTransform: jest.SpyInstance

function initMocks() {
  beforeEach(() => {
    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    pxTransform.mockRestore()
  })
}

export default initMocks