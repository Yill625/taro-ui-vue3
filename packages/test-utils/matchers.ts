import * as utils from "@taro-ui-vue3/utils/common"

jest.mock('@taro-ui-vue3/utils/common')

let warn: jest.SpyInstance
let uuid: jest.SpyInstance

const asserted: Set<string> = new Set()
const NOOP = () => { }

// Ensure console.warn is called with expected message
const toHaveBeenWarned: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed = warn.mock.calls.some(args => args[0].indexOf(received) > -1)
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned` +
        (msgs.length
          ? `.\n\nActual messages:\n\n - ${msgs}`
          : ` but no warning was recorded.`)
    }
  }
}

// Ensure previous console.warn is called with expected message
const toHaveBeenWarnedLast: jest.CustomMatcher = (received: string) => {
  asserted.add(received)
  const passed =
    warn.mock.calls[warn.mock.calls.length - 1][0].indexOf(received) > -1
  if (passed) {
    return {
      pass: true,
      message: () => `expected "${received}" not to have been warned last.`
    }
  } else {
    const msgs = warn.mock.calls.map(args => args[0]).join('\n - ')
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned last.\n\nActual messages:\n\n - ${msgs}`
    }
  }
}

// Ensure console.warn is called n times with expected message
const toHaveBeenWarnedTimes: jest.CustomMatcher = (received: string, n: number) => {
  asserted.add(received)
  let found = 0
  warn.mock.calls.forEach(args => {
    if (args[0].indexOf(received) > -1) {
      found++
    }
  })

  if (found === n) {
    return {
      pass: true,
      message: () => `expected "${received}" to have been warned ${n} times.`
    }
  } else {
    return {
      pass: false,
      message: () =>
        `expected "${received}" to have been warned ${n} times but got ${found}.`
    }
  }
}

function initWarnMatchers() {
  beforeEach(() => {
    asserted.clear()
    warn = jest.spyOn(console, 'warn').mockImplementation(NOOP)
    uuid = jest.spyOn(utils, 'uuid').mockReturnValue('__tests__2021')

    // NOT working in concurrent mode
    // should avoid use in test.concurrent()
    expect.extend({
      toHaveBeenWarned,
      toHaveBeenWarnedLast,
      toHaveBeenWarnedTimes
    })
  })

  afterEach(() => {
    const assertedArray = Array.from(asserted)
    const nonAssertedWarnings = warn.mock.calls
      .map(args => args[0])
      .filter(received => {
        return !assertedArray.some(assertedMsg => {
          return received.indexOf(assertedMsg) > -1
        })
      })

    uuid.mockRestore()
    warn.mockRestore()

    if (nonAssertedWarnings.length) {
      throw new Error(
        `test case threw unexpected warnings:\n - ${nonAssertedWarnings.join(
          '\n - '
        )}`
      )
    }
  })
}

export default initWarnMatchers