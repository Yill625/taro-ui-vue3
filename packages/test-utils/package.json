{
  "name": "@taro-ui-vue3/test-utils",
  "version": "0.0.0",
  "main": "dist/index.js",
  "license": "MIT",
  "dependencies": {
    "vue": "^3.2.0",
    "@vue/test-utils": "^2.0.0-rc.0"
  }
}