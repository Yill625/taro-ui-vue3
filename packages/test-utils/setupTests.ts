import toHaveBeenWarnedInit from './been-warned'

toHaveBeenWarnedInit()