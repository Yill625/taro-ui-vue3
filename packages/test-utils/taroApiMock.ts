import * as utils from "@taro-ui-vue3/utils/common"
import Taro from '@tarojs/taro'
import { createAnimation } from '@tarojs/taro-h5'

import {
  getEnvMockFn,
  pxTransformMockFn,
  chooseImageMockFn,
  querySelectorMockFn
} from './@tarojs/taro'

Taro.pxTransform = pxTransformMockFn
Taro.chooseImage = chooseImageMockFn
Taro.createSelectorQuery = querySelectorMockFn()
Taro.createAnimation = createAnimation

let uuid: jest.SpyInstance
let getEnvs: jest.SpyInstance
let getenv: jest.SpyInstance
let pxTransform: jest.SpyInstance
let chooseImage: jest.SpyInstance

function initTaroMockApis() {
  beforeEach(() => {
    jest.mock('@tarojs/taro', () => {
      return Taro
    })

    jest.mock('@taro-ui-vue3/utils/common', () => {
      return utils
    })

    getenv = jest
      .spyOn(Taro, 'getEnv')
      .mockImplementation(getEnvMockFn)

    chooseImage = jest
      .spyOn(Taro, 'chooseImage')
      .mockImplementation(chooseImageMockFn)

    pxTransform = jest
      .spyOn(utils, 'pxTransform')
      .mockImplementation(pxTransformMockFn)

    uuid = jest
      .spyOn(utils, 'uuid')
      .mockReturnValue('__tests__2021')

    getEnvs = jest
      .spyOn(utils, 'getEnvs')
      .mockImplementation(() => {
        const env = getEnvMockFn()
        return {
          isWEAPP: env === 'WEAPP',
          isALIPAY: env === 'ALIPAY',
          isWEB: env === 'WEB'
        }
      })
  })

  afterEach(() => {
    uuid.mockRestore()
    getEnvs.mockRestore()
    getenv.mockRestore()
    pxTransform.mockRestore()
    chooseImage.mockRestore()
  })
}

export default initTaroMockApis
