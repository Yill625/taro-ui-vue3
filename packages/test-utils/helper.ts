import { mount, shallowMount, DOMWrapper, VueWrapper } from '@vue/test-utils'
import { VNode, Slot } from '@vue/runtime-core'

export const sleep = async (timeout: number): Promise<null> =>
  new Promise((resolve) => setTimeout(resolve, timeout))

export interface Slots {
  default: Array<string | VNode> | Slot
  [x: string]: Array<string | VNode> | Slot
}

export function mountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return mount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export function genMountFn(
  comp: any,
  components?: Record<string, object>,
  useShallowMount?: boolean
) {
  return (props = {}, slots?: Slots) => {
    return useShallowMount
      ? shallowMountFactory(comp, components, props, slots)
      : mountFactory(comp, components, props, slots)
  }
}

export function shallowMountFactory(
  comp: any,
  components?: Record<string, object>,
  props?: Record<string, any>,
  slots?: Slots
) {
  return shallowMount<any>(comp, {
    global: {
      components
    },
    props,
    slots,
  })
}

export interface Dimension {
  top?: number
  bottom?: number
  left?: number
  right?: number
  width?: number
  height?: number
}

export function genDelayedSelectorSpy(
  utils: any,
  dimensions: Dimension
): jest.SpyInstance {
  return jest
    .spyOn(utils, 'delayQuerySelector')
    .mockImplementation((_, sel: string, delay?: number) => {
      return new Promise((resolve) => {
        resolve([dimensions])
      })
    })
}

export async function triggerTouchEvents(
  el: DOMWrapper<Element>,
  touchstartDetails: any,
  touchendDetails: any
) {
  const startDetails = {
    touches: [touchstartDetails],
    changedTouches: [touchstartDetails]
  }

  const endDetails = {
    touches: [touchendDetails],
    changedTouches: [touchendDetails]
  }

  await el.trigger('touchstart', startDetails)
  await el.trigger('touchmove', endDetails)
  await el.trigger('touchend', endDetails)
}

export function testPropClassAndStyle(
  mountFn: (props?: {}, slots?: Slots | undefined) => VueWrapper<any>
) {

  it.concurrent('should render prop class', async () => {
    const wrapper = mountFn({ class: 'test' })
    expect(
      wrapper.classes()
    ).toContain('test')
  })

  it.concurrent('should render prop style', async () => {
    const wrapper = mountFn({ style: { color: 'red' } })
    expect(
      wrapper.attributes('style')
    ).toContain('color: red;')
  })
}

export function hexToRGBA(hex: string, opacity = 1) {
  hex = hex.replace('#', '')
  const rgb = hex
    .match(new RegExp('(.{' + hex.length / 3 + '})', 'g'))!
    .map((l) => parseInt(hex.length % 2 ? l + l : l, 16))

  const rgba = rgb.concat(isFinite(opacity) ? opacity : 1)

  return opacity === 1
    ? `rgb(${rgb.join(', ')})`
    : `rgba(${rgba.join(', ')})`
}