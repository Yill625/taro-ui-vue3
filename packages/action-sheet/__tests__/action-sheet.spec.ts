import { mountFactory, Slots } from '@taro-ui-vue3/test-utils/helper'
import { h } from '@vue/runtime-core'
import AtActionSheet from '../index'
import AtActionSheetItem from '../body/item'

const factory = (
  values = {},
  slots: Slots = {
    default: []
  }
) => {
  return mountFactory(AtActionSheet, { AtActionSheetItem }, values, slots)
}

describe('ActionSheet', () => {
  it('should render default ActionSheet', () => {
    const wrapper = factory()
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render opened ActionSheet', () => {
    const wrapper = factory({
      isOpened: true,
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render opened ActionSheet with slot contents', () => {
    const wrapper = factory({
      isOpened: true,
    }, {
      default: [
        h(AtActionSheetItem, null, { default: () => '按钮一' }),
        h(AtActionSheetItem, null, { default: () => '按钮二' })
      ]
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render opened ActionSheet with prop cancelText', () => {
    const wrapper = factory({
      isOpened: true,
      cancelText: '取消',
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find('.at-action-sheet__footer').exists()).toBe(true)
  })

  it('should not render ActionSheetFooter without props cancelText', () => {
    const wrapper = factory({
      isOpened: true
    })
    expect(wrapper.find('.at-action-sheet__footer').exists()).toBe(false)
  })

  it('should render opened ActionSheet -- props title', () => {
    const wrapper = factory({
      isOpened: true,
      title: '清除位置信息后， 别人将不能查看到你\r\n可以通过转义字符换行',
    })
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.find('.at-action-sheet__header').exists()).toBe(true)
  })

  it('should not render ActionSheetHeader without prop title', () => {
    const wrapper = factory({
      isOpened: true
    })
    expect(wrapper.find('.at-action-sheet__header').exists()).toBe(false)
  })

  it('should render opened ActionSheet with header, body and footer', () => {
    const wrapper = factory({
      isOpened: true,
      cancelText: '取消',
      title: '清除位置信息后， 别人将不能查看到你\r\n可以通过转义字符换行',
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})

describe('ActionSheet events', () => {

  it.concurrent('should trigger onCancel & onClose events', async () => {
    const onCancel = jest.fn()
    const onClose = jest.fn()

    const wrapper = factory(
      {
        isOpened: true,
        cancelText: '取消',
        title: '清除位置信息后， 别人将不能查看到你\r\n可以通过转义字符换行',
        onCancel: onCancel,
        onClose: onClose,
      },
      {
        default: [
          h(AtActionSheetItem, null, { default: () => '按钮一' })
        ],
      }
    )

    wrapper.find('.at-action-sheet__footer').trigger('tap')
    wrapper.find('.at-action-sheet__overlay').trigger('tap')
    expect(onCancel).toBeCalled()
    expect(onClose).toBeCalled()
  })

  it.concurrent('should trigger onClose event if prop open changed to false', async () => {
    const onClose = jest.fn()

    const wrapper = factory(
      {
        isOpened: true,
        cancelText: '取消',
        title: '清除位置信息后， 别人将不能查看到你\r\n可以通过转义字符换行',
        onClose: onClose,
      },
      {
        default: [
          h(AtActionSheetItem, null, { default: () => '按钮一' })
        ],
      }
    )

    await wrapper.setProps({ isOpened: false })
    expect(onClose).toBeCalled()
  })

  it.concurrent('should trigger onClose event if prop open changed to true', async () => {
    const onClose = jest.fn()

    const wrapper = factory(
      {
        cancelText: '取消',
        title: '清除位置信息后， 别人将不能查看到你\r\n可以通过转义字符换行',
        onClose: onClose,
      },
      {
        default: [
          h(AtActionSheetItem, null, { default: () => '按钮一' })
        ],
      }
    )

    await wrapper.setProps({ isOpened: true })
    expect(onClose).not.toBeCalled()
  })

  it.concurrent('should trigger onClick event on AtActionSheetItem', async () => {
    const onClick = jest.fn()

    const wrapper = factory({
      isOpened: true,
      cancelText: '取消',
      title: '清除位置信息后， 别人将不能查看到你\r\n可以通过转义字符换行'
    }, {
      default: [
        h(AtActionSheetItem, { onClick: onClick }, { default: () => '按钮一' }),
        h(AtActionSheetItem, null, { default: () => '按钮二' })
      ]
    })

    wrapper.findAll('.at-action-sheet__item').forEach(el => {
      el.trigger('tap')
    })
    expect(onClick.mock.calls.length).toBe(1)
  })

})
