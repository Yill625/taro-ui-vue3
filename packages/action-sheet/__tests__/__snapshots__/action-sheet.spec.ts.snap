// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionSheet should render default ActionSheet 1`] = `
<view
  class="at-action-sheet"
>
  <view
    class="at-action-sheet__overlay"
  />
  <view
    class="at-action-sheet__container"
  >
    
    <view
      class="at-action-sheet__body"
    />
    
  </view>
</view>
`;

exports[`ActionSheet should render opened ActionSheet -- props title 1`] = `
<view
  class="at-action-sheet at-action-sheet--active"
>
  <view
    class="at-action-sheet__overlay"
  />
  <view
    class="at-action-sheet__container"
  >
    <view
      class="at-action-sheet__header"
    >
      清除位置信息后， 别人将不能查看到你
可以通过转义字符换行
    </view>
    <view
      class="at-action-sheet__body"
    />
    
  </view>
</view>
`;

exports[`ActionSheet should render opened ActionSheet 1`] = `
<view
  class="at-action-sheet at-action-sheet--active"
>
  <view
    class="at-action-sheet__overlay"
  />
  <view
    class="at-action-sheet__container"
  >
    
    <view
      class="at-action-sheet__body"
    />
    
  </view>
</view>
`;

exports[`ActionSheet should render opened ActionSheet with header, body and footer 1`] = `
<view
  class="at-action-sheet at-action-sheet--active"
>
  <view
    class="at-action-sheet__overlay"
  />
  <view
    class="at-action-sheet__container"
  >
    <view
      class="at-action-sheet__header"
    >
      清除位置信息后， 别人将不能查看到你
可以通过转义字符换行
    </view>
    <view
      class="at-action-sheet__body"
    />
    <view
      class="at-action-sheet__footer"
    >
      取消
    </view>
  </view>
</view>
`;

exports[`ActionSheet should render opened ActionSheet with prop cancelText 1`] = `
<view
  class="at-action-sheet at-action-sheet--active"
>
  <view
    class="at-action-sheet__overlay"
  />
  <view
    class="at-action-sheet__container"
  >
    
    <view
      class="at-action-sheet__body"
    />
    <view
      class="at-action-sheet__footer"
    >
      取消
    </view>
  </view>
</view>
`;

exports[`ActionSheet should render opened ActionSheet with slot contents 1`] = `
<view
  class="at-action-sheet at-action-sheet--active"
>
  <view
    class="at-action-sheet__overlay"
  />
  <view
    class="at-action-sheet__container"
  >
    
    <view
      class="at-action-sheet__body"
    >
      <view
        class="at-action-sheet__item"
      >
        按钮一
      </view>
      <view
        class="at-action-sheet__item"
      >
        按钮二
      </view>
    </view>
    
  </view>
</view>
`;
